### DISCUSSION TITLE
Why does anyExtendedKeyUsage have a value of 85.29.37.0 and not 2.5.29.37.0?

### DISCUSSION THREAD

**Discussion by wzshiming** [openssl/openssl#23051] [Q&A] [answered]
Use curl to call a server using server certificates generated by another SSL library with anyExtendedKeyUsage.

Curl version
```
curl 8.5.0 (aarch64-apple-darwin23.0.0) libcurl/8.5.0 (SecureTransport) OpenSSL/3.2.0 zlib/1.2.12 brotli/1.1.0 zstd/1.5.5 libidn2/2.3.4 libssh2/1.11.0 nghttp2/1.58.0 librtmp/2.3 OpenLDAP/2.6.6
```

Error message
```
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unsupported certificate (555):
* SSL certificate problem: unsuitable certificate purpose
* Closing connection
curl: (60) SSL certificate problem: unsuitable certificate purpose
More details here: https://curl.se/docs/sslcerts.html
```

I parse it from the constant definition and it looks like `85.29.37.0`
https://github.com/openssl/openssl/blob/5056133cc7d6c52033c25b2e1f7762bcafcce760/crypto/objects/obj_dat.h#L861
https://github.com/openssl/openssl/blob/5056133cc7d6c52033c25b2e1f7762bcafcce760/fuzz/oids.txt#L857

I searched the others and they all seem to be `2.5.29.37.0`
https://github.com/golang/go/blob/5f48662c5aab0f8c594b77ca183687264121f6b5/src/crypto/x509/x509.go#L600
http://oidref.com/2.5.29.37.0

Did I get something wrong?

**Comment by t8m** [comment, marked as answer]
These binary data values are in the encoded form. The encoded OID is actually the correct one `2.5.29.37.0`.

**Reply by davidben** [reply]
Specifically the first two components get combined as 40\*A + B when encoding to save a byte. 40\*2 + 5 is 85, hence the first byte.

**Comment by wzshiming** [comment]
@t8m

Sorry, I have one more question

``` 
$ openssl x509 -in admin.crt -noout -purpose
Certificate purposes:
SSL client : No
SSL client CA : No
SSL server : No
SSL server CA : No
Netscape SSL server : No
Netscape SSL server CA : No
S/MIME signing : No
S/MIME signing CA : No
S/MIME encryption : No
S/MIME encryption CA : No
CRL signing : No
CRL signing CA : No
Any Purpose : Yes
Any Purpose CA : Yes
OCSP helper : Yes
OCSP helper CA : No
Time Stamp signing : No
Time Stamp signing CA : No
Code signing : No
Code signing CA : No
```

My server certificate has this `Any Purpose : Yes`, but why is curl reporting an error?

```
curl -vv --cacert ca.crt https://127.0.0.1:8080
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /path/to/ca.crt
*  CApath: none
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unsupported certificate (555):
* SSL certificate problem: unsuitable certificate purpose
* Closing connection
curl: (60) SSL certificate problem: unsuitable certificate purpose
More details here: https://curl.se/docs/sslcerts.html

```

**Reply by tom-cosgrove-arm** [reply]
@wzshiming Could I suggest starting a new discussion for a different problem, since this discussion is titled "Why does anyExtendedKeyUsage have a value of 85.29.37.0 and not 2.5.29.37.0?" and that question has been answered?
